proposals/2020-07-01_aws-kms-keyring-redesign/background.md

```diff
@@ -194,83 +211,77 @@ The AWS KMS symmetric region discovery keyring MAY...
-   that passes an [AWS KMS Decrypt API](https://docs.aws.amazon.com/kms/latest/APIReference/API_Decrypt.html)
-   authorization check for the encrypted data key.
-2. The keyring is unable to write a decryption contract,
-   but it MAY fulfill any AWS KMS decryption contract for the configured AWS region and, if provided, the AWS account ID.
-   1. Fulfilling any AWS KMS decryption contract MUST be sufficient to succeed.
+Depending on the implementation, keyring-producing operations can be thought of as factory methods/builders.
+They initialize keyrings.
 
-**Initialization**
+**Please note the following applies to the keyring-producing operations
+that produce the [derived keyrings](#derived-keyrings) below.**
 
-On keyring initialization, an AWS KMS symmetric region discovery keyring MUST define the following:
+**All keyring-producing operations MUST...**
 
-- AWS SDK client (and the AWS SDK client’s region)
+1. Produce a single [derived keyring](#derived-keyrings)
+   1. Each keyring-producing operation MUST be configured with the additional arguments or parameters
+      required to satisfy the specification of the derived keyring it produces.
+2. Initialize the AWS SDK client(s) required for communicating with AWS KMS for the given key names/AWS regions
+3. Configure each AWS KMS keyring with the AWS SDK client it needs to communicate with AWS KMS
+4. Be configurable with an optional AWS SDK client configuration (client config) that includes custom AWS SDK credentials
+   or a similar AWS SDK language-specific client configuration option
+   (examples include:
+   Java's [Client Configuration](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/ClientConfiguration.html)
+   AND [AWS Credentials Provider](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/auth/AWSCredentialsProvider.html),
+   Python's [botocore Config](https://boto3.amazonaws.com/v1/documentation/api/latest/guide/configuration.html)
+   AND [botocore Session](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html),
+   etc.)
+   1. If a client config is provided,
+      the keyring-producing operation MUST configure all AWS SDK clients with the provided client config
+5. Limit the initialization of new AWS SDK client(s) when possible
+   1. Multiple keyrings can be configured with the same AWS SDK client if the following conditions are all met:
+      1. Keyrings MUST be configured in the same keyring-producing operation call
+      2. Keyrings MUST communicate with AWS KMS in the same AWS region
+      3. AWS SDK clients MUST share the same AWS SDK client configuration
+6. Maintain the order of all key names/AWS regions provided by the customer
 
-On keyring initialization, an AWS KMS symmetric region discovery keyring MAY define the following:
+**All keyring-producing operations MAY...**
 
-- AWS account ID
-- Grant Tokens
+1. Add an identifier to the user agent string in the AWS SDK client
```

Comment:
This only applies to the keyring-producing operations for now, since that's where we construct the AWS SDK client. I believe in most languages, the client is directly associated with a Config that contains the user agent string, so we don't update this again once the client is created.

During the implementation phase, we'll determine if there are languages were the user agent string is instead passed per call instead of per client. If the string is passed when the call is made and cannot be passed directly to the client for all calls, we might need to amend this behavior a bit.